bodhi/tests/server/services/test_updates.py

```diff
@@ -5167,3 +5167,82 @@ def test_unexpected_exception(self, log_exception, waive_test_results, *args):
+
+        up = self.db.query(Update).filter_by(title=nvr).one()
+        up.locked = False
+
+        post_data = dict(update=nvr,
+                         csrf_token=self.app.get('/csrf').json_body['csrf_token'])
+        res = self.app.post_json('/updates/%s/get-test-results' % str(nvr), post_data, status=400)
+
+        self.assertEquals(res.json_body['status'], 'error')
+        self.assertEquals(res.json_body[u'errors'][0][u'description'],
+                          "No greenwave_api_url specified")
+
+    @unittest.skipIf(six.PY3, 'Not working with Python 3 yet')
+    @mock.patch('bodhi.server.services.updates.Update.get_test_gating_info',
+                side_effect=IOError('IOError. oops!'))
+    @mock.patch('bodhi.server.services.updates.log.exception')
+    def test_unexpected_exception(self, log_exception, get_test_gating_info, *args):
+        """Ensure that an unexpected Exception is handled by get_test_results()."""
+        nvr = u'bodhi-2.0-1.fc17'
+
+        up = self.db.query(Update).filter_by(title=nvr).one()
+        up.locked = False
+
+        post_data = dict(update=nvr, request='stable',
+                         csrf_token=self.app.get('/csrf').json_body['csrf_token'])
+        res = self.app.post_json('/updates/%s/get-test-results' % str(nvr), post_data, status=400)
+
+        self.assertEquals(res.json_body['status'], 'error')
+        self.assertEquals(res.json_body['errors'][0]['description'],
+                          u'IOError. oops!')
+        log_exception.assert_called_once_with("Unhandled exception in get_test_results")
+
+    @unittest.skipIf(six.PY3, 'Not working with Python 3 yet')
+    @mock.patch.dict(config, [('greenwave_api_url', 'https://greenwave.api')])
+    @mock.patch('bodhi.server.util.call_api')
+    def test_get_test_results_calling_greenwave(self, call_api, *args):
+        """
+        Ensure if all conditions are met we do try to call greenwave with the proper
+        argument.
+        """
+        nvr = u'bodhi-2.0-1.fc17'
+        call_api.return_value = True
+
+        up = self.db.query(Update).filter_by(title=nvr).one()
+        up.locked = False
+
+        post_data = dict(update=nvr,
+                         csrf_token=self.app.get('/csrf').json_body['csrf_token'])
+        self.app.post_json('/updates/%s/get-test-results' % str(nvr), post_data)
+
+        call_api.assert_called_once_with(
+            'https://greenwave.api/decision',
+            data={
+                'product_version': u'fedora-17',
+                'decision_context': u'bodhi_update_push_testing',
+                'subject': [
+                    {'item': u'bodhi-2.0-1.fc17', 'type': 'koji_build'},
+                    {'original_spec_nvr': u'bodhi-2.0-1.fc17'},
+                    {'item': u'FEDORA-2018-a3bbe1a8f2', 'type': 'bodhi_update'}
+                ]
+            },
+            method='POST',
+            retries=3,
+            service_name='Greenwave'
+        )
```

Comment:
That would mean querying greenwave during the tests no?

Reply:
We can use the mock to set the return value to some particular dictionary, like {'return': 'value'} and then just assert that the right data is returned. Otherwise there isn't a test that proves that the API actually does what it says it does.